### ISSUE TITLE
instance predicate returns false for objects

### ISSUE THREAD

**Issue by philbooth** [philbooth/check-types.js#52]
Raised [elsewhere](https://github.com/philbooth/check-types.js/issues/38#issuecomment-135220534) by @euoia:

>I have found a case where check.instance doesn't agree with instanceof. In Chrome and Safari:
>```
>document.getElementById('el') instanceof HTMLElement
>> true
>```
>
>But due to how the instance function in check-types.js checks that prototype is a function (see https://github.com/philbooth/check-types.js/blob/master/src/check-types.js#L151), the following returns false in Safari and true in Chrome:
>```
>check.assert.instance(document.getElementById('el'), HTMLElement);
>```
>
>Due to this behaviour in Safari:
>```
>typeof HTMLElement
>> "object"
>```
>
>And this in Chrome:
>```
>typeof HTMLElement
>> "function"
>```
>
>So although this seems due to a browser implementation difference, it is at least surprising that check.instance doesn't agree with instanceof. Is it necessary for the instance function to check that prototype is a function?

**Comment by philbooth** [comment]
Yeah, whatever the reason for doing that was, I don't remember it now. If there even was a reason. :confused:

Will fix it.

**Comment by philbooth** [comment]
Oh wait, I do know why it's there.

If you use a non-function as the rhs to `instanceof`, it throws a `TypeError`. At least it does for me in Firefox, Chrome and node.js. What happens when you do `null instanceof {}` in your Safari console?

**Comment by euoia** [comment]
It throws a `TypeError`:
```
null instanceof {}
> TypeError: Object is not a valid argument for 'instanceof' (evaluating 'null instanceof {}')
```

**Comment by philbooth** [comment]
Thanks. I'm minded to keep the current behaviour at the moment, or perhaps a modified version of it.

I made a conscious decision a few versions back that the predicates should never throw, they must always be relied on to return `true` or `false`. Before then, the `like` predicate had a condition in which it threw for instance, which made calling code more complex than it needed to be. So I'd definitely like to adhere to the 'never throw from a predicate' principle.

But maybe it is possible to special-case the `instance` predicate so that it handles your example sanely without causing errors to be thrown when random objects get passed to it. I haven't thought about how to do that, so feel free to make suggestions if you have any.

In the meantime, you could work around it using assert-as-a-function:

```js
check.assert(document.getElementById('el') instanceof HTMLElement);
```

What do you think?

**Comment by euoia** [comment]
Oh, I didn't realise you can do that. That's useful, thanks.

I completely agree that predicates should never throw. It seems like for now special casing instance might be the best solution.

Thanks again for taking the time to respond to this.

**Comment by philbooth** [comment]
Decided to fix this just by catching the error. I'm mystified why that didn't occur to me before tbh. :confused:

Anyway, published on npm as 4.0.0, [with a new unit test added for this case](https://github.com/philbooth/check-types.js/blob/master/test/check-types.js#L113-L117).
